### PR TITLE
Allow passing through custom http client in IO socket connect

### PR DESCRIPTION
**Pull request by Rexios80**
I am using a service that is very strict about user agents. Passing a user agent into the `headers` field doesn't work because this code adds headers instead of replacing them.

https://github.com/dart-lang/sdk/blob/83b7f7c2ac2df0a375387dc1017f6c96ccbcaccc/sdk/lib/_http/websocket_impl.dart#L1029

This PR allows passing in a custom http client to resolve the issue

### CODE REVIEW

**Comment by Rexios80** [comment]
@kevmoo Sorry for the ping, but is this package actively maintained? I see other similar issues and PRs with little to no activity. I need this feature merged for use in one of my own published packages, so using a git or path dependency isn't an option. I really don't want to release my own version of this package, but I'll have to if we can't get this merged relatively soon.

**Comment by kevmoo** [comment]
@brianquinlan ?

**Comment by kevmoo** [comment]
@Rexios80 – we should have tests here!

**Comment by Rexios80** [comment]
Formatting fixed and test added

**Comment by kevmoo** [comment]
Or maybe @devoncarew / @natebosch ?

File: lib/io.dart [hunk 1]
```diff
@@ -74,13 +74,15 @@ class IOWebSocketChannel extends StreamChannelMixin
...
     Duration? connectTimeout,
+    HttpClient? customClient,
   }) {
     late IOWebSocketChannel channel;
     final sinkCompleter = WebSocketSinkCompleter();
     var future = WebSocket.connect(
       url.toString(),
       headers: headers,
       protocols: protocols,
+      customClient: customClient,
```
Comment (natebosch) [inline lib/io.dart:85]: How about calling it `client`. I don't think the word "custom" is adding value

**Comment by natebosch** [comment]
Seems like a fine change. Will need CHANGELOG

File: lib/io.dart (hunk 1 above)
Comment (Rexios80) [reply lib/io.dart:85]: I copied the parameter name from here:
Screenshot 2023-04-06 at 3 00 56 PM

Are we sure we want to change it?

**Comment by Rexios80** [comment]
Those tests pass for me locally 🤔

File: lib/io.dart (hunk 1 above)
Comment (natebosch) [reply lib/io.dart:85]: Ah good call. `customClient` SGTM.

**Comment by Rexios80** [comment]
Oh... I guess the `customClient` parameter didn't exist in Dart `2.12.0`? Should we care?

**Comment by Rexios80** [comment]
Turns out that parameter didn't exist until Dart 2.15.0. I increased the minimum Dart SDK version and updated the action to use it.

**Comment by natebosch** [comment]
I'll fix the existing diagnostics in https://github.com/dart-lang/web_socket_channel/pull/262 and then CI should be green here.

**Review by natebosch — APPROVED** [review]

**Comment by Rexios80** [comment]
@natebosch Do you have an ETA on releasing this to pub.dev?

**Comment by natebosch** [comment]
This will likely get published Monday, it needs to first roll into the sdk and internal to Google

**Comment by Rexios80** [comment]
Relevant PRs:
https://github.com/dart-lang/web_socket_channel/pull/215
https://github.com/dart-lang/web_socket_channel/pull/211

Relevant issues (there might be more):
https://github.com/dart-lang/web_socket_channel/issues/86

These should be closed since this PR addressed them

**Comment by natebosch** [comment]
This hasn't landed internally yet. I'll watch for when it is rolled and publish after than.

**Comment by natebosch** [comment]
https://pub.dev/packages/web_socket_channel/versions/2.4.0

**Comment by Rexios80** [comment]
@natebosch Do you want to close those other PRs/issues I mentioned?

**Pull request merged**